crates/handlers/src/admin/v1/oauth2_sessions/finish.rs

```diff
@@ -0,0 +1,234 @@
+    AlreadyFinished(Ulid),
+}
+
+impl_from_error_for_route!(mas_storage::RepositoryError);
+
+impl IntoResponse for RouteError {
+    fn into_response(self) -> axum::response::Response {
+        let error = ErrorResponse::from_error(&self);
+        let sentry_event_id = record_error!(self, Self::Internal(_));
+        let status = match self {
+            Self::Internal(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            Self::NotFound(_) => StatusCode::NOT_FOUND,
+            Self::AlreadyFinished(_) => StatusCode::BAD_REQUEST,
+        };
+        (status, sentry_event_id, Json(error)).into_response()
+    }
+}
+
+pub fn doc(operation: TransformOperation) -> TransformOperation {
+    operation
+        .id("finishOAuth2Session")
+        .summary("Finish an OAuth 2.0 session")
+        .description(
+            "Calling this endpoint will finish the OAuth 2.0 session, preventing any further use. If the session has a user associated with it, a job will be scheduled to sync the user's devices with the homeserver.",
+        )
+        .tag("oauth2-session")
+        .response_with::<200, Json<SingleResponse<OAuth2Session>>, _>(|t| {
+            // Get the finished session sample
+            let [_, _, finished_session] = OAuth2Session::samples();
+            let id = finished_session.id();
+            let response = SingleResponse::new(
+                finished_session,
+                format!("/api/admin/v1/oauth2-sessions/{id}/finish"),
+            );
+            t.description("OAuth 2.0 session was finished").example(response)
+        })
+        .response_with::<400, RouteError, _>(|t| {
+            let response = ErrorResponse::from_error(&RouteError::AlreadyFinished(Ulid::nil()));
+            t.description("Session is already finished")
+                .example(response)
+        })
+        .response_with::<404, RouteError, _>(|t| {
+            let response = ErrorResponse::from_error(&RouteError::NotFound(Ulid::nil()));
+            t.description("OAuth 2.0 session was not found")
+                .example(response)
+        })
+}
+
+#[tracing::instrument(name = "handler.admin.v1.oauth2_sessions.finish", skip_all)]
+pub async fn handler(
+    CallContext {
+        mut repo, clock, ..
+    }: CallContext,
+    NoApi(mut rng): NoApi<BoxRng>,
+    id: UlidPathParam,
+) -> Result<Json<SingleResponse<OAuth2Session>>, RouteError> {
+    let id = *id;
+    let session = repo
+        .oauth2_session()
+        .lookup(id)
+        .await?
+        .ok_or(RouteError::NotFound(id))?;
+
+    // Check if the session is already finished
+    if session.finished_at().is_some() {
+        return Err(RouteError::AlreadyFinished(id));
+    }
+
+    // If the session has a user associated with it, schedule a job to sync devices
+    if let Some(user_id) = session.user_id {
+        tracing::info!(user.id = %user_id, "Scheduling device sync job for user");
+        let job = SyncDevicesJob::new_for_id(user_id);
```

Comment:
weirdly inconsistent with how it's done for CompatSessions; why use the db-loaded `User` in one place and the ID in the other?

I think just using the `user_id` should win, though; should be no need to verify the user exists in the DB given the FK constraint should do that for us

Reply:
Good q, dunno why I did that on the compat sessions. Fixed that in 7e3df5585